closed/src/java.base/share/classes/openj9/internal/security/RestrictedSecurity.java

```diff
@@ -0,0 +1,1053 @@
+                    securityProps.getProperty(propsPrefix + ".tls.ephemeralDHKeySize"));
+            jdkTlsLegacyAlgorithms = parseProperty(
+                    securityProps.getProperty(propsPrefix + ".tls.legacyAlgorithms"));
+            jdkCertpathDisabledAlgorithms = parseProperty(
+                    securityProps.getProperty(propsPrefix + ".jce.certpath.disabledAlgorithms"));
+            jdkSecurityLegacyAlgorithm = parseProperty(
+                    securityProps.getProperty(propsPrefix + ".jce.legacyAlgorithms"));
+            keyStoreType = parseProperty(
+                    securityProps.getProperty(propsPrefix + ".keystore.type"));
+            keyStore = parseProperty(
+                    securityProps.getProperty(propsPrefix + ".javax.net.ssl.keyStore"));
+
+            jdkSecureRandomProvider = parseProperty(
+                    securityProps.getProperty(propsPrefix + ".securerandom.provider"));
+            jdkSecureRandomAlgorithm = parseProperty(
+                    securityProps.getProperty(propsPrefix + ".securerandom.algorithm"));
+
+            if (debug != null) {
+                debug.println("Loaded restricted security properties.");
+            }
+        }
+
+        /**
+         * Load security constraints with type, algorithm, attributes.
+         *
+         * Example:
+         * RestrictedSecurity1.jce.provider.1 = SUN [{CertPathBuilder, PKIX, *},
+         * {Policy, JavaPolicy, *}, {CertPathValidator, *, *}].
+         */
+        private void initConstraints() {
+            for (int pNum = 1; pNum <= providersSimpleName.size(); pNum++) {
+                String providerName = providersSimpleName.get(pNum - 1);
+                String providerInfo = securityProps
+                        .getProperty(propsPrefix + ".jce.provider." + pNum);
+
+                if (debug != null) {
+                    debug.println("Loading constraints for security provider: " + providerName);
+                }
+
+                // Check if the provider has constraints
+                if (providerInfo.indexOf('[') < 0 && providerInfo.indexOf(']') < 0) {
+                    if (debug != null) {
+                        debug.println("No constraints for security provider: " + providerName);
+                    }
+                    providerConstraints.put(providerName, new Constraint[0]);
+                    continue;
+                }
+
+                // Remove the whitespaces in the format separator if there are.
+                providerInfo = providerInfo.trim()
+                        .replaceAll("\\[\\s*\\{", "[{")
+                        .replaceAll("\\}\\s*\\]", "}]")
+                        .replaceAll("\\}\\s*,\\s*\\{", "},{");
+
+                int startIndex = providerInfo.indexOf("[{");
+                int endIndex = providerInfo.indexOf("}]");
+
+                // Provider with constraints.
+                if ((startIndex > 0) && (endIndex > 0)) {
```

Comment:
This should ensure that there's only one `[{` and one `}]` that occur in that order.
```java
                int startIndex = providerInfo.lastIndexOf("[{");
                int endIndex = providerInfo.indexOf("}]");

                // Provider with constraints.
                if ((startIndex > 0) && (endIndex > startIndex)) {
```
(`areBracketsBalanced()` would accept `"[{ ... }], [{ ... }]"`)